### PR TITLE
Added summary of error handling and HTTP status codes

### PR DESCRIPTION
**Pull request by buske**
PR to resolve issue #24.  Comments, edits, and suggestions welcome.

### CODE REVIEW

File: search-api.md [hunk 1]
```diff
@@ -230,3 +224,26 @@ A synchronous `application/json` response, of the following form:
...
 * ***Mandatory***, but can be empty
 * Is a **list of matches**, where each match has a `patient` object of the same format as the one described above for the query
+
+
+### Error handling
+The remote server should use HTTP status codes to report any errors encoundered processing the match request. Here are a list of status codes and their meanings with regards to this API:
+
+| HTTP Status Code | Constant | Description
+| ---------------- | -------- | -----------
+| 200 | httplib.OK | no error |
```
Comment (sdumitriu) [inline search-api.md]: Remove `httplib.` since it's a Python specific module, and not all matchmakers are written in Python.

Comment (allasm) [inline search-api.md]: can probably replace by names as in http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html (which are basically the same without the prefix)

File: search-api.md [hunk 2]
```diff
@@ -230,3 +224,26 @@ A synchronous `application/json` response, of the following form:
 #### Results
 * ***Mandatory***, but can be empty
 * Is a **list of matches**, where each match has a `patient` object of the same format as the one described above for the query
+
+
+### Error handling
+The remote server should use HTTP status codes to report any errors encoundered processing the match request. Here are a list of status codes and their meanings with regards to this API:
+
+| HTTP Status Code | Constant | Description
```
Comment (sdumitriu) [inline search-api.md]: I would replace `Constant` with `Reason Phrase`, and switch the values to Title Case, [as in the RFC](http://tools.ietf.org/html/rfc7231#section-6).

File: search-api.md [hunk 3]
```diff
@@ -230,3 +224,26 @@ A synchronous `application/json` response, of the following form:
...
+| ---------------- | -------- | -----------
+| 200 | httplib.OK | no error |
+| 400 | httplib.BAD_REQUEST | missing/invalid data
+| 401 | httplib.UNAUTHORIZED | invalid API key
+| 405 | httplib.METHOD_NOT_ALLOWED | invalid method (GET)
+| 406 | httplib.NOT_ACCEPTABLE | unsupported API version
+| 415 | httplib.UNSUPPORTED_MEDIA_TYPE | missing/invalid content type
+| 422 | httplib.UNPROCESSABLE_ENTITY | missing/invalid request body
+| 500 | httplib.INTERNAL_SERVER_ERROR | default error
+
```
Comment (sdumitriu) [inline search-api.md]: Missing words: "providing details [?] the error"

File: search-api.md [hunk 4]
```diff
@@ -230,3 +224,26 @@ A synchronous `application/json` response, of the following form:
...
+The error response should include a json-formatted body with a human-readable `"message"` providing details the error. For example, if the match request specifies an unsupported API version, the server should respond with `Not Acceptable (406)` and content such as:
+
+```
+{
+  "message" : "unsupported version number"
```
Comment (allasm) [inline search-api.md]: Unless we want to specify all messages we should explicitly note that the exact error message may be up to the implementer? I also suggest to allow implementations to add other fields as they see fit, e.g. ""supportedVersions": [ "1.0", "1.1", "2.0" ] - just for developer convenience

File: search-api.md [hunk 5]
```diff
@@ -23,12 +23,6 @@ After receiving a request, the remote server can respond in one of two ways:
   * If a compatible version (`vX.Z` where `Z>=Y`) is supported by the remote server, it should provide a response using this version.
   * If no appropriate version is supported by the remote server, it should respond with `Not Acceptable (406)`, containing a JSON body with a description of the error. All responses, including this one, should contain a `Content-Type` header with the latest API version supported by the server. This will enable the user to re-submit the request using this version of the API.
```
Comment (allasm) [inline search-api.md]: Hmm, say a client submits a request using version 1.0, which *is* supported by the server. Shouldn't the server reply using v1.0 specifications (even if the latest supported version is 2.0 and is totally different and incompatible?). If that is the case, Content-Type of the reply should match the actual content, and not the latest supported version. I don't want to start a versioning hell and fine to require any server to support only one version :), but if multiple versions are supported content-type should be consistent with actual content.

If we want to add ability to discover the list of versions (or latest version) supported by the server I suggest to add a list of supported versions to the "unsupported version number" JSON reply (and possibly specify that version "0.0" should never be supported and can be used - in the unlikely case if anyone ever wants this - to get a guaranteed incorrect version reply with the list of supported versions?

**Comment by buske** [comment]
Thanks for the feedback. I updated the pull request accordingly.

File: search-api.md (hunk 5 above)
Comment (buske) [inline search-api.md]: Hmm, I'm a bit confused. If the client submits a request using v1.0, which *is* supported, then the server should respond with v1.0 (as per the first bullet point). If it isn't supported, I think the server should respond with an error with the Content-Type of the latest version supported by the server. This gives the requester a version to resubmit with, if they so choose. It's not full-fledged version negotiation, but seemed like a reasonable first-pass that provides 99% of the functionality we'd need.

Comment (allasm) [inline search-api.md]: OK, I see the confusion point - I assumed that the phrase "All responses, including this one, should contain a `Content-Type` header with the latest API version supported by the server" includes the case of replies in supported versions other than the latest. If this only applies to unsupported versions I would rephrase it as "All error responses...."

Comment (sdumitriu) [inline search-api.md]: What @allasm is complaining about is this part: 

> All responses, including this one, should contain a `Content-Type` header with the latest API version supported by the server.

This does conflict a bit with the previous line:

> If a compatible version [...] is supported by the remote server, it should provide a response using this version.

How do we define _compatible_? Are bugfix `X.*` versions supposed to be compatible, with a clearly defined measure of [in]compatibility? Should the exact same version as in the request be used, or do we consider that all versions sharing the same major number are compatible and a client supporting one version should be able to handle (possibly ignoring certain fields) any other version from the same branch?

Comment (buske) [inline search-api.md]: I see now what you're complaining about @allasm. I'll try to clarify that now. 

@sdumitriu, I defined compatible in that line as same major version and at least that minor version (`vX.Z` is a compatible response to `vX.Y` where `Z>=Y`). Then, any minor releases should only include additional fields (or remove optional fields), which can then be ignored in a backwards-compatible way.

Comment (buske) [inline search-api.md]: Is that reasonable?

**Comment by sdumitriu** [comment]
Looks good, +1.

File: search-api.md [hunk 6]
```diff
@@ -230,3 +224,27 @@ A synchronous `application/json` response, of the following form:
...
+| 415 | Unsupported Media Type | missing/invalid content type
+| 422 | Unprocessable Entity | missing/invalid request body
+| 500 | Internal Server Error | default error
+
+The error response should include a json-formatted body with a human-readable `"message"` containing further details about the error. The exact error message is up to the implementer, and additional fields can be provided with further information. For example, if the match request specifies an unsupported API version, the server should respond with `Not Acceptable (406)` and a content body such as:
+
+```
+{
+  "message" : "unsupported API version",
+  "supportedVersions" : [ "0.1", "1.0", "1.1" ]
```
Comment (Relequestual) [inline search-api.md]: I was wondering if we should instead return the supported versions in an accept header, but that might not be the right solution.

**Comment by Relequestual** [comment]
I'm happy. +1

**Comment by fschiettecatte** [comment]
+1

**Comment by Relequestual** [comment]
That's three. Merging and closing.
